### PR TITLE
Fail the explore when one or more hardware exploration fails.

### PR DESCRIPTION
**Pull request by kema2000**

### CODE REVIEW

File: src/test/kotlin/com/atlassian/performance/tools/hardware/HardwareExploration.kt [hunk 1]
```diff
@@ -70,7 +71,11 @@ class HardwareExploration(
         val awsExecutor = Executors.newFixedThreadPool(awsParallelism)
         val explorationExecutor = Executors.newFixedThreadPool(space.size)
         try {
-            return exploreHardwareInParallel(space, explorationExecutor, awsExecutor)
+            val result = exploreHardwareInParallel(space, explorationExecutor, awsExecutor)
+            if(!hardwaresWithError.isEmpty()){
```
Comment (dagguh) [inline src/test/kotlin/com/atlassian/performance/tools/hardware/HardwareExploration.kt]: Code style - I recommend reformatting via IntelliJ all the time

Comment (dagguh) [inline src/test/kotlin/com/atlassian/performance/tools/hardware/HardwareExploration.kt]: `!hardwaresWithError.isEmpty()` -> `hardwaresWithError.isNotEmpty()`

File: src/test/kotlin/com/atlassian/performance/tools/hardware/HardwareExploration.kt [hunk 2]
```diff
@@ -60,6 +60,7 @@ class HardwareExploration(
 ) {
     private val awsParallelism = 6
     private val results = ConcurrentHashMap<Hardware, Future<HardwareExplorationResult>>()
+    private val hardwaresWithError = mutableSetOf<Hardware>()
```
Comment (dagguh) [inline src/test/kotlin/com/atlassian/performance/tools/hardware/HardwareExploration.kt]:
```suggestion
    private val failedHardware = mutableSetOf<Hardware>()
```

File: src/test/kotlin/com/atlassian/performance/tools/hardware/HardwareExploration.kt [hunk 3]
```diff
@@ -70,7 +71,11 @@ class HardwareExploration(
         val awsExecutor = Executors.newFixedThreadPool(awsParallelism)
         val explorationExecutor = Executors.newFixedThreadPool(space.size)
         try {
-            return exploreHardwareInParallel(space, explorationExecutor, awsExecutor)
+            val result = exploreHardwareInParallel(space, explorationExecutor, awsExecutor)
+            if(!hardwaresWithError.isEmpty()){
+                throw Exception("One or more tests failed : $hardwaresWithError ")
```
Comment (dagguh) [inline src/test/kotlin/com/atlassian/performance/tools/hardware/HardwareExploration.kt]: No need for the trailing space

Comment (dagguh) [inline src/test/kotlin/com/atlassian/performance/tools/hardware/HardwareExploration.kt]: We should refer to QUICK-216 in the commit message

**Review by dagguh — APPROVED** [review]

**Comment by dagguh** [comment]
I raised some small nit-picks, nice-to-haves, which are entirely optional

Good job. I like how non-intrusive and local this fix is 👍

File: src/test/kotlin/com/atlassian/performance/tools/hardware/HardwareExploration.kt (hunk 1 above)
Comment (kema2000) [reply src/test/kotlin/com/atlassian/performance/tools/hardware/HardwareExploration.kt]: sure

**Review by mminns — APPROVED** [review]

**Pull request merged**